File: app/src/main/java/com/chartapp/Adapter.java
package com.chartapp;

import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class Adapter extends RecyclerView.Adapter<Adapter.MyViewHolder> implements Filterable {
        ArrayList<Data> objData;
        Intent intent;

private List<Data> exampleListFull;


public Adapter(ArrayList<Data> objData) {
        this.objData = objData;
        exampleListFull = new ArrayList<>(objData);

        }


@NonNull
@Override
public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View singleRow = LayoutInflater.from(parent.getContext()).inflate(R.layout.list,parent,false);
        return new MyViewHolder(singleRow);
        }

@Override
public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        Data objData = this.objData.get(position);
        holder.name.setText(objData.getName());
        holder.other.setText(objData.getOther());
        holder.name.setOnClickListener(new View.OnClickListener() {
@Override
public void onClick(View v) {
        intent = new Intent(Intent.ACTION_VIEW, Uri.parse(objData.getRef()));
        v.getContext().startActivity(Adapter.this.intent);

        }
        });

        }

@Override
public int getItemCount() {
        return objData == null ? 0 : objData.size();
        }

public static class MyViewHolder extends RecyclerView.ViewHolder{
    TextView name;
    TextView other;



    public MyViewHolder(@NonNull View itemView) {
        super(itemView);
        name = itemView.findViewById(R.id.name);
        other = itemView.findViewById(R.id.other);


    }
}
    public Filter getFilter() {
        return exampleFilter;
    }

    private Filter exampleFilter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            List<Data> filteredList = new ArrayList<>();

            if (constraint == null || constraint.length() == 0) {
                filteredList.addAll(exampleListFull);
            } else {
                String filterPattern = constraint.toString().toLowerCase().trim();

                for (Data item : exampleListFull) {
                    if (item.getName().toLowerCase().contains(filterPattern)) {
                        filteredList.add(item);
                    }
                }
            }

            FilterResults results = new FilterResults();
            results.values = filteredList;

            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            objData.clear();
            objData.addAll((List)results.values);
            notifyDataSetChanged();
        }
    };

}

File: app/src/main/java/com/chartapp/Adapter.kt
package com.chartapp

import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.chartapp.Adapter.MyViewHolder
import java.util.*

class Adapter(objData: ArrayList<Data>) : RecyclerView.Adapter<MyViewHolder>(), Filterable {
    var objData: ArrayList<Data>?
    var intent: Intent? = null
    private val exampleListFull: List<Data>

    init {
        this.objData = objData
        exampleListFull = ArrayList(objData)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val singleRow = LayoutInflater.from(parent.context).inflate(R.layout.list, parent, false)
        return MyViewHolder(singleRow)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        val objData = objData!![position]
        holder.name.text = objData.name
        holder.other.text = objData.other
        holder.name.setOnClickListener { v ->
            intent = Intent(Intent.ACTION_VIEW, Uri.parse(objData.ref))
            v.context.startActivity(intent)
        }
    }

    override fun getItemCount(): Int {
        return if (objData == null) 0 else objData!!.size
    }

    class MyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var name: TextView
        var other: TextView

        init {
            name = itemView.findViewById(R.id.name)
            other = itemView.findViewById(R.id.other)
        }
    }

    override fun getFilter(): Filter {
        return exampleFilter
    }

    private val exampleFilter: Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val filteredList: MutableList<Data> = ArrayList()
            if (constraint == null || constraint.length == 0) {
                filteredList.addAll(exampleListFull)
            } else {
                val filterPattern =
                    constraint.toString().lowercase(Locale.getDefault()).trim { it <= ' ' }
                for (item in exampleListFull) {
                    if (item.name.lowercase(Locale.getDefault()).contains(filterPattern)) {
                        filteredList.add(item)
                    }
                }
            }
            val results = FilterResults()
            results.values = filteredList
            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            objData.clear()
            objData.addAll(results.values as Collection<Data>)
            notifyDataSetChanged()
        }
    }


}